src/libstd/sys_common/parking_lot_core/word_lock.rs

```diff
@@ -0,0 +1,271 @@
+    #[inline]
+    pub fn new() -> WordLock {
+        WordLock {
+            state: AtomicUsize::new(0),
+        }
+    }
+
+    #[inline]
+    pub unsafe fn lock(&self) {
+        if self
+            .state
+            .compare_exchange_weak(0, LOCKED_BIT, Ordering::Acquire, Ordering::Relaxed)
+            .is_ok()
+        {
+            return;
+        }
+        self.lock_slow();
+    }
+
+    #[inline]
+    pub unsafe fn unlock(&self) {
+        let state = self.state.fetch_sub(LOCKED_BIT, Ordering::Release);
+        if state & QUEUE_LOCKED_BIT != 0 || state & QUEUE_MASK == 0 {
+            return;
+        }
+        self.unlock_slow();
+    }
+
+    #[cold]
+    #[inline(never)]
+    unsafe fn lock_slow(&self) {
+        let mut spinwait = SpinWait::new();
+        let mut state = self.state.load(Ordering::Relaxed);
+        loop {
+            // Grab the lock if it isn't locked, even if there is a queue on it
+            if state & LOCKED_BIT == 0 {
+                match self.state.compare_exchange_weak(
+                    state,
+                    state | LOCKED_BIT,
+                    Ordering::Acquire,
+                    Ordering::Relaxed,
+                ) {
+                    Ok(_) => return,
+                    Err(x) => state = x,
+                }
+                continue;
+            }
+
+            // If there is no queue, try spinning a few times
+            if state & QUEUE_MASK == 0 && spinwait.spin() {
+                state = self.state.load(Ordering::Relaxed);
+                continue;
+            }
+
+            // Get our thread data and prepare it for parking
+            let mut thread_data = None;
+            let thread_data = get_thread_data(&mut thread_data);
+            assert!(mem::align_of_val(thread_data) > !QUEUE_MASK);
+            thread_data.parker.prepare_park();
+
+            // Add our thread to the front of the queue
+            let queue_head = (state & QUEUE_MASK) as *const ThreadData;
+            if queue_head.is_null() {
+                thread_data.queue_tail.set(thread_data);
+                thread_data.prev.set(ptr::null());
+            } else {
+                thread_data.queue_tail.set(ptr::null());
+                thread_data.prev.set(ptr::null());
+                thread_data.next.set(queue_head);
+            }
+            if let Err(x) = self.state.compare_exchange_weak(
+                state,
+                (state & !QUEUE_MASK) | thread_data as *const _ as usize,
+                Ordering::Release,
+                Ordering::Relaxed,
+            ) {
+                state = x;
+                continue;
+            }
+
+            // Sleep until we are woken up by an unlock
+            thread_data.parker.park();
+
+            // Loop back and try locking again
+            spinwait.reset();
+            self.state.load(Ordering::Relaxed);
+        }
+    }
+
+    #[cold]
+    #[inline(never)]
+    unsafe fn unlock_slow(&self) {
```

Comment:
This is a lot of logic for an unsafe function and the invariants aren't stated;
It would be more readable to split it up into smaller chunks imo and also state the invariants required.
(Applies also to the function before this one).

Reply:
In this case the invariants are pretty simple: don't lock if the current thread already owns the lock and don't unlock if the current thread doesn't own the lock.

Reply:
@Amanieu Sure; but it should be stated in documentation explicitly. :)

Reply:
I have now managed to make three methods in this type not fully `unsafe`. And the only `unsafe` one left states why it is: `unlock()` must not be called on an already unlocked `WordLock`.